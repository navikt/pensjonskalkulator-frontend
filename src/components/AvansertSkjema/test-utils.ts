export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
