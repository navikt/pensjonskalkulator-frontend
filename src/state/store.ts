import {
  combineReducers,
  configureStore,
  createListenerMiddleware,
  ListenerEffectAPI,
  TypedStartListening,
  TypedAddListener,
  PreloadedState,
} from '@reduxjs/toolkit'

import { apiSlice } from './api/apiSlice'
import { createUttaksalderListener } from './listeners/uttaksalderListener'
import userInputReducer from './userInput/userInputReducer'

const listenerMiddleware = createListenerMiddleware()

export const rootReducer = combineReducers({
  [apiSlice.reducerPath]: apiSlice.reducer,
  userInput: userInputReducer,
})

export function setupStore(
  preloadedState?: PreloadedState<RootState>,
  isDev = false
) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware(
        isDev ? { immutableCheck: false, serializableCheck: false } : {}
      )
        .concat(apiSlice.middleware)
        .prepend(listenerMiddleware.middleware),
  })
}

export const store = setupStore()

createUttaksalderListener(
  listenerMiddleware.startListening as AppStartListening
)

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']

export type AppListenerEffectAPI = ListenerEffectAPI<RootState, AppDispatch>
export type AppStartListening = TypedStartListening<RootState, AppDispatch>
export type AppAddListener = TypedAddListener<RootState, AppDispatch>
