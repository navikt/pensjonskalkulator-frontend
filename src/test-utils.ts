import renderer from 'react-test-renderer'

import { render } from '@testing-library/react'

export function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => children,
    ...options,
  })

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
