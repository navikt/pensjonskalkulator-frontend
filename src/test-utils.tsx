import React, { PropsWithChildren } from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import { PreloadedState, createListenerMiddleware } from '@reduxjs/toolkit'
import { render, RenderOptions } from '@testing-library/react'

import { getTranslation_test } from '@/utils/__tests__/test-translations'

import { createUttaksalderListener } from './state/listeners/uttaksalderListener'
import {
  setupStore,
  RootState,
  AppStore,
  AppStartListening,
} from './state/store'
import { getTranslation_nb } from './translations/nb'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  hasRouter?: boolean
}

export const swallowErrors = (testFn: () => void) => {
  const error = console.error
  console.error = () => {}
  testFn()
  console.error = error
}

export const swallowErrorsAsync = async (testFn: () => Promise<void>) => {
  const cache = console.error
  console.error = () => {}
  await testFn()
  console.error = cache
}

function generateMockedTranslations() {
  const nbTranslations: Record<string, string> = getTranslation_nb()
  const testTranslations: Record<string, string> = getTranslation_test()
  const translationsInput = { ...nbTranslations, ...testTranslations }
  const translations: Record<string, string> = {}

  for (const key in translationsInput) {
    if (
      /<(?=.*? .*?\/ ?>|br|hr|input|!--|wbr)[a-z]+.*?>|<([a-z]+).*?<\/\1>|{[^}]+}/i.test(
        translationsInput[key]
      )
    ) {
      // for html or chunks keys: results in 'my_key' : 'my_key with some <html> or {chunk}'
      translations[key] = translationsInput[key]
    } else {
      // results in 'my_key' : 'my_key'
      translations[key] = key
    }
  }
  return translations
}

// Return an object with the store and all of RTL's query functions
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState, true),
    hasRouter = true,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <IntlProvider locale="nb" messages={generateMockedTranslations()}>
          {hasRouter ? <MemoryRouter>{children}</MemoryRouter> : children}
        </IntlProvider>
      </Provider>
    )
  }

  const listenerMiddleware = createListenerMiddleware()
  createUttaksalderListener(
    listenerMiddleware.startListening as AppStartListening
  )

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'

export { renderWithProviders as render }
