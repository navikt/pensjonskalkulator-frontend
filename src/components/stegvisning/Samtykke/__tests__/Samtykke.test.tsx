import { describe, it, vi } from 'vitest'

import { Samtykke } from '..'
import { RootState } from '@/state/store'
import { screen, render, waitFor, userEvent } from '@/test-utils'

describe('stegvisning - Samtykke', () => {
  const onCancelMock = vi.fn()
  const onPreviousMock = vi.fn()
  const onNextMock = vi.fn()
  it('rendrer slik den skal når samtykket ikke er oppgitt', async () => {
    const result = render(
      <Samtykke
        harSamtykket={null}
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />
    )
    expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(
      'stegvisning.samtykke.title'
    )
    const radioButtons = screen.getAllByRole('radio')

    await waitFor(() => {
      expect(screen.getAllByRole('radio')).toHaveLength(2)
      expect(radioButtons[0]).not.toBeChecked()
      expect(radioButtons[1]).not.toBeChecked()
      expect(result.asFragment()).toMatchSnapshot()
    })
  })
  describe('rendrer slik den skal når samtykket er oppgitt', async () => {
    it('når samtykket er true', async () => {
      render(
        <Samtykke
          harSamtykket
          onCancel={onCancelMock}
          onPrevious={onPreviousMock}
          onNext={onNextMock}
        />
      )
      const radioButtons = screen.getAllByRole('radio')
      await waitFor(() => {
        expect(screen.getAllByRole('radio')).toHaveLength(2)
        expect(radioButtons[0]).toBeChecked()
        expect(radioButtons[1]).not.toBeChecked()
      })
    })

    it('når samtykket er false', async () => {
      render(
        <Samtykke
          harSamtykket={false}
          onCancel={onCancelMock}
          onPrevious={onPreviousMock}
          onNext={onNextMock}
        />
      )
      const radioButtons = screen.getAllByRole('radio')
      await waitFor(() => {
        expect(screen.getAllByRole('radio')).toHaveLength(2)
        expect(radioButtons[0]).not.toBeChecked()
        expect(radioButtons[1]).toBeChecked()
      })
    })
  })

  it('validerer, viser feilmelding, fjerner feilmelding og kaller onNext når brukeren klikker på Neste', async () => {
    const user = userEvent.setup()
    render(
      <Samtykke
        harSamtykket={null}
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />
    )
    const radioButtons = screen.getAllByRole('radio')

    await user.click(screen.getByText('stegvisning.neste'))

    waitFor(() => {
      expect(
        screen.getByText('stegvisning.samtykke.validation_error')
      ).toBeInTheDocument()
      expect(onNextMock).not.toHaveBeenCalled()
    })

    await user.click(radioButtons[0])

    expect(
      screen.queryByText('stegvisning.samtykke.validation_error')
    ).not.toBeInTheDocument()

    await user.click(screen.getByText('stegvisning.neste'))

    waitFor(() => {
      expect(onNextMock).toHaveBeenCalled()
    })
  })

  it('kaller onPrevious når brukeren klikker på Tilbake', async () => {
    const user = userEvent.setup()
    render(
      <Samtykke
        harSamtykket
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />,
      {
        preloadedState: { userInput: { samtykke: true } } as RootState,
      }
    )
    const radioButtons = screen.getAllByRole('radio')
    expect(radioButtons[0]).toBeChecked()

    await user.click(screen.getByText('stegvisning.tilbake'))

    waitFor(() => {
      expect(onPreviousMock).toHaveBeenCalled()
    })
  })

  it('kaller onCancelMock når brukeren klikker på Avbryt', async () => {
    const user = userEvent.setup()
    render(
      <Samtykke
        harSamtykket
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />
    )

    await user.click(screen.getByText('stegvisning.avbryt'))

    waitFor(() => {
      expect(onCancelMock).toHaveBeenCalled()
    })
  })
  it('viser avbryt knapp når onCancel er definert', async () => {
    render(
      <Samtykke
        harSamtykket
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />
    )

    expect(screen.getByText('stegvisning.avbryt')).toBeInTheDocument()
  })

  it('viser ikke avbryt knapp når onCancel ikke er definert', async () => {
    render(
      <Samtykke
        harSamtykket
        onCancel={undefined}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />
    )
    expect(screen.queryByText('stegvisning.avbryt')).not.toBeInTheDocument()
  })
})
