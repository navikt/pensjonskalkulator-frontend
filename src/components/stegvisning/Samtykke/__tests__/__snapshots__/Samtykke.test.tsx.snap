// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stegvisning - Samtykke > rendrer slik den skal n√•r samtykket ikke er oppgitt 1`] = `
<DocumentFragment>
  <section
    class="_section_67454f _section__largePadding_67454f _section__marginBotton_67454f"
  >
    <form>
      <h2
        class="navds-heading navds-heading--medium navds-typo--spacing"
      >
        stegvisning.samtykke.title
      </h2>
      <p
        class="navds-body-long navds-body-long--large"
      >
        stegvisning.samtykke.ingress
      </p>
      <div
        class="navds-read-more navds-read-more--medium _readmore_5b4738"
      >
        <button
          aria-expanded="false"
          class="navds-read-more__button navds-body-short"
          data-testid="readmore"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="navds-read-more__expand-icon"
            fill="none"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M5.97 9.47a.75.75 0 0 1 1.06 0L12 14.44l4.97-4.97a.75.75 0 1 1 1.06 1.06l-5.5 5.5a.75.75 0 0 1-1.06 0l-5.5-5.5a.75.75 0 0 1 0-1.06"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <span>
            stegvisning.samtykke.readmore_title
          </span>
        </button>
        <div
          aria-hidden="true"
          class="navds-read-more__content navds-read-more__content--closed navds-body-long navds-body-long--medium"
        >
          stegvisning.samtykke.readmore_list_title
          <ul
            class="_list_5b4738"
          >
            <li>
              stegvisning.samtykke.readmore_list_item1
            </li>
            <li>
              stegvisning.samtykke.readmore_list_item2
            </li>
            <li>
              stegvisning.samtykke.readmore_list_item3
            </li>
          </ul>
          stegvisning.samtykke.readmore_ingress
        </div>
      </div>
      <fieldset
        aria-required="true"
        class="_radiogroup_5b4738 navds-radio-group navds-radio-group--medium navds-fieldset navds-fieldset--medium"
        role="radiogroup"
      >
        <legend
          class="navds-fieldset__legend navds-label"
        >
          stegvisning.samtykke.radio_label
        </legend>
        <div
          class="navds-radio-buttons"
        >
          <div
            class="navds-radio navds-radio--medium"
          >
            <input
              aria-labelledby="r4"
              class="navds-radio__input"
              id="radio-r3"
              name="samtykke"
              type="radio"
              value="ja"
            />
            <label
              class="navds-radio__label"
              for="radio-r3"
            >
              <span
                class="navds-radio__content"
              >
                <span
                  aria-hidden="true"
                  class="navds-body-short navds-body-short--medium"
                  id="r4"
                >
                  stegvisning.samtykke.radio_ja
                </span>
              </span>
            </label>
          </div>
          <div
            class="navds-radio navds-radio--medium"
          >
            <input
              aria-labelledby="r7"
              class="navds-radio__input"
              id="radio-r6"
              name="samtykke"
              type="radio"
              value="nei"
            />
            <label
              class="navds-radio__label"
              for="radio-r6"
            >
              <span
                class="navds-radio__content"
              >
                <span
                  aria-hidden="true"
                  class="navds-body-short navds-body-short--medium"
                  id="r7"
                >
                  stegvisning.samtykke.radio_nei
                </span>
              </span>
            </label>
          </div>
        </div>
        <div
          aria-live="polite"
          aria-relevant="additions removals"
          class="navds-fieldset__error"
          id="fieldset-error-r2"
        />
      </fieldset>
      <button
        class="_button_5b4738 navds-button navds-button--primary navds-button--medium"
        type="submit"
      >
        <span
          class="navds-label"
        >
          stegvisning.neste
        </span>
      </button>
      <button
        class="_button_5b4738 navds-button navds-button--secondary navds-button--medium"
        type="button"
      >
        <span
          class="navds-label"
        >
          stegvisning.tilbake
        </span>
      </button>
      <button
        class="_button_5b4738 navds-button navds-button--tertiary navds-button--medium"
        type="button"
      >
        <span
          class="navds-label"
        >
          stegvisning.avbryt
        </span>
      </button>
    </form>
  </section>
</DocumentFragment>
`;
